promql/test.go

```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
+	m2, m2hash := getHistogramMetricBase(m, suffix)
+	histWrapper, exists := histMap[m2hash]
+	if !exists {
+		histWrapper = newTempHistogramWrapper()
+	}
+	histWrapper.metric = m2
+	if updateHistWrapper != nil {
+		updateHistWrapper(&histWrapper)
+	}
+	for _, s := range smpls {
+		if s.H != nil {
+			continue
+		}
+		hist, exists := histWrapper.histByTs[s.T]
+		if !exists {
+			hist = newTempHistogram()
+		}
+		updateHist(&hist, s.F)
+		histWrapper.histByTs[s.T] = hist
+	}
+	histMap[m2hash] = histWrapper
+}
+
+// if classic histograms are defined, convert them into native histograms with custom
+// bounds and append the defined time series to the storage.
+func (cmd *loadCmd) appendCustomHistogram(a storage.Appender) error {
+	histMap := map[uint64]tempHistogramWrapper{}
+
+	// Go through all the time series to collate classic histogram data
+	// and organise them by timestamp.
+	for hash, smpls := range cmd.defs {
+		m := cmd.metrics[hash]
+		mName := m.Get(labels.MetricName)
+		switch {
+		case strings.HasSuffix(mName, "_bucket") && m.Has(labels.BucketLabel):
+			le, err := strconv.ParseFloat(m.Get(labels.BucketLabel), 64)
+			if err != nil {
+				continue
+			}
+			processClassicHistogramSeries(m, "_bucket", histMap, smpls, func(histWrapper *tempHistogramWrapper) {
+				histWrapper.upperBounds = append(histWrapper.upperBounds, le)
+			}, func(hist *tempHistogram, f float64) {
+				hist.bucketCounts[le] = f
+			})
+		case strings.HasSuffix(mName, "_count"):
+			processClassicHistogramSeries(m, "_count", histMap, smpls, nil, func(hist *tempHistogram, f float64) {
+				hist.count = f
+			})
+		case strings.HasSuffix(mName, "_sum"):
+			processClassicHistogramSeries(m, "_sum", histMap, smpls, nil, func(hist *tempHistogram, f float64) {
+				hist.sum = f
+			})
+		}
+	}
+
+	// Convert the collated classic histogram data into native histograms
+	// with custom bounds and append them to the storage.
+	for _, histWrapper := range histMap {
+		sort.Float64s(histWrapper.upperBounds)
+		upperBounds := make([]float64, 0, len(histWrapper.upperBounds))
+		prevLe := -372.4869 // assuming that the lowest bound is never this value
+		for _, le := range histWrapper.upperBounds {
+			if le != prevLe { // deduplicate
+				upperBounds = append(upperBounds, le)
+				prevLe = le
+			}
+		}
+		var customBounds []float64
+		if upperBounds[len(upperBounds)-1] == math.Inf(1) {
+			customBounds = upperBounds[:len(upperBounds)-1]
+		} else {
+			customBounds = upperBounds
+		}
+		fhBase := &histogram.FloatHistogram{
+			Count:  0,
+			Sum:    0,
+			Schema: histogram.CustomBucketsSchema,
+			PositiveSpans: []histogram.Span{
+				{Offset: 0, Length: uint32(len(upperBounds))},
+			},
+			CustomValues: customBounds,
+		}
+		ts := make([]int64, 0, len(histWrapper.histByTs))
+		for t := range histWrapper.histByTs {
+			ts = append(ts, t)
+		}
+		sort.Slice(ts, func(i, j int) bool { return ts[i] < ts[j] })
```

Comment:
I think it would be simpler to just create the slice of `Sample`s  unsorted , call sort.Slice on that, before calling appendSample. So this pre-sorting into `ts` can be removed altogether.